File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/LinqExpr.ts
// 對應 C# 的 LinqExpr 相關類別與介面
import { ComparisonOperator } from './ComparisonOperator';
import { LogicalOperator } from './LogicalOperator';

export class LinqExpr {
    From!: LinqFromExpr;
    Where?: LinqWhereExpr;
    OrderBy?: LinqOrderByExpr;
    Joins?: LinqJoinExpr[];
    AdditionalFroms?: LinqFromExpr[];
    Select?: ILinqExpression;
}

export class LinqFromExpr {
    Source!: string;
    AliasName!: string;
    IsDefaultIfEmpty!: boolean;
}

export class LinqSelectAllExpr implements ILinqExpression {
    AliasName!: string;
    equals(obj: any): boolean {
        return obj instanceof LinqSelectAllExpr && obj.AliasName === this.AliasName;
    }
}

export class LinqWhereExpr {
    Condition!: ILinqExpression;
}

export interface ILinqExpression {
    equals(obj: any): boolean;
}

export class LinqConditionExpression implements ILinqExpression {
    Left!: ILinqExpression;
    ComparisonOperator!: ComparisonOperator;
    Right!: ILinqExpression;
    LogicalOperator?: LogicalOperator;
    equals(obj: any): boolean {
        if (!(obj instanceof LinqConditionExpression)) return false;
        if (this.LogicalOperator || obj.LogicalOperator) {
            return this.Left.equals(obj.Left) && this.Right.equals(obj.Right) && this.LogicalOperator === obj.LogicalOperator;
        }
        return this.Left.equals(obj.Left) && this.ComparisonOperator === obj.ComparisonOperator && this.Right.equals(obj.Right);
    }
}

export class LinqFieldExpr implements ILinqExpression {
    TableOrAlias?: string;
    FieldName: string = '';
    equals(obj: any): boolean {
        return obj instanceof LinqFieldExpr && obj.TableOrAlias === this.TableOrAlias && obj.FieldName === this.FieldName;
    }
}

export class LinqValue implements ILinqExpression {
    Value: string = '';
    equals(obj: any): boolean {
        return obj instanceof LinqValue && obj.Value === this.Value;
    }
}

export class LinqOrderByExpr implements ILinqExpression {
    Fields: LinqOrderByFieldExpr[] = [];
    equals(obj: any): boolean {
        return obj instanceof LinqOrderByExpr && JSON.stringify(this.Fields) === JSON.stringify(obj.Fields);
    }
}

export class LinqOrderByFieldExpr implements ILinqExpression {
    Field!: LinqFieldExpr;
    IsDescending!: boolean;
    equals(obj: any): boolean {
        return obj instanceof LinqOrderByFieldExpr && this.Field.equals(obj.Field) && this.IsDescending === obj.IsDescending;
    }
}

export class LinqJoinExpr implements ILinqExpression {
    JoinType: string = 'join';
    AliasName!: string;
    Source!: string;
    On!: LinqConditionExpression;
    Into?: string;
    equals(obj: any): boolean {
        return obj instanceof LinqJoinExpr && this.JoinType === obj.JoinType && this.AliasName === obj.AliasName && this.Source === obj.Source && this.On.equals(obj.On) && this.Into === obj.Into;
    }
}

export class LinqSelectNewExpr implements ILinqExpression {
    Fields!: LinqSelectFieldExpr[];
    equals(obj: any): boolean {
        return obj instanceof LinqSelectNewExpr && JSON.stringify(this.Fields) === JSON.stringify(obj.Fields);
    }
}

export class LinqSelectFieldExpr implements ILinqExpression {
    Name!: string;
    Value!: ILinqExpression;
    equals(obj: any): boolean {
        return obj instanceof LinqSelectFieldExpr && this.Name === obj.Name && this.Value.equals(obj.Value);
    }
} 

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/LinqExpr.cs
namespace T1.SqlSharp.Expressions;

public class LinqExpr
{
    public LinqFromExpr From { get; set; }
    public LinqWhereExpr? Where { get; set; }
    public LinqOrderByExpr? OrderBy { get; set; }
    public List<LinqJoinExpr>? Joins { get; set; }
    public List<LinqFromExpr>? AdditionalFroms { get; set; }
    public ILinqExpression? Select { get; set; }
}

public class LinqFromExpr
{
    public string Source { get; set; }
    public string AliasName { get; set; }
    public bool IsDefaultIfEmpty { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqFromExpr other) return false;
        return Source == other.Source && AliasName == other.AliasName && IsDefaultIfEmpty == other.IsDefaultIfEmpty;
    }
    public override int GetHashCode() => (Source, AliasName, IsDefaultIfEmpty).GetHashCode();
}

public class LinqSelectAllExpr : ILinqExpression
{
    public string AliasName { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqSelectAllExpr other) return false;
        return AliasName == other.AliasName;
    }
    public override int GetHashCode() => AliasName.GetHashCode();
}

public class LinqWhereExpr
{
    public ILinqExpression Condition { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqWhereExpr other) return false;
        return Equals(Condition, other.Condition);
    }
    public override int GetHashCode() => Condition?.GetHashCode() ?? 0;
}

public interface ILinqExpression
{
    bool Equals(object? obj);
    int GetHashCode();
}

public class LinqConditionExpression : ILinqExpression
{
    public ILinqExpression Left { get; set; }
    public ComparisonOperator ComparisonOperator { get; set; }
    public ILinqExpression Right { get; set; }
    public LogicalOperator? LogicalOperator { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqConditionExpression other) return false;
        if (LogicalOperator.HasValue || other.LogicalOperator.HasValue)
        {
            return Equals(Left, other.Left) && Equals(Right, other.Right) && Nullable.Equals(LogicalOperator, other.LogicalOperator);
        }
        return Equals(Left, other.Left) && ComparisonOperator == other.ComparisonOperator && Equals(Right, other.Right);
    }
    public override int GetHashCode() => (Left, ComparisonOperator, Right, LogicalOperator).GetHashCode();
}

public class LinqFieldExpr : ILinqExpression
{
    public string? TableOrAlias { get; set; }
    public string FieldName { get; set; } = string.Empty;
    public override bool Equals(object? obj)
    {
        if (obj is not LinqFieldExpr other) return false;
        return TableOrAlias == other.TableOrAlias && FieldName == other.FieldName;
    }
    public override int GetHashCode() => (TableOrAlias, FieldName).GetHashCode();
}

public class LinqValue : ILinqExpression
{
    public string Value { get; set; } = string.Empty;
    public override bool Equals(object? obj)
    {
        if (obj is not LinqValue other) return false;
        return Value == other.Value;
    }
    public override int GetHashCode() => Value.GetHashCode();
}

public class LinqOrderByExpr : ILinqExpression
{
    public List<LinqOrderByFieldExpr> Fields { get; set; } = new();
    public override bool Equals(object? obj)
    {
        if (obj is not LinqOrderByExpr other) return false;
        return Fields.SequenceEqual(other.Fields);
    }
    public override int GetHashCode() => Fields.GetHashCode();
}

public class LinqOrderByFieldExpr : ILinqExpression
{
    public LinqFieldExpr Field { get; set; }
    public bool IsDescending { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqOrderByFieldExpr other) return false;
        return Equals(Field, other.Field) && IsDescending == other.IsDescending;
    }
    public override int GetHashCode() => (Field, IsDescending).GetHashCode();
}

public class LinqJoinExpr : ILinqExpression
{
    public string JoinType { get; set; } = "join";
    public string AliasName { get; set; }
    public string Source { get; set; }
    public LinqConditionExpression On { get; set; }
    public string? Into { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqJoinExpr other) return false;
        return JoinType == other.JoinType
            && AliasName == other.AliasName
            && Source == other.Source
            && Equals(On, other.On)
            && Into == other.Into;
    }
    public override int GetHashCode() => (JoinType, AliasName, Source, On, Into).GetHashCode();
}

public class LinqSelectNewExpr : ILinqExpression
{
    public List<LinqSelectFieldExpr> Fields { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqSelectNewExpr other) return false;
        return Fields.SequenceEqual(other.Fields);
    }
    public override int GetHashCode() => Fields.GetHashCode();
}

public class LinqSelectFieldExpr : ILinqExpression
{
    public string Name { get; set; }
    public ILinqExpression Value { get; set; }
    public override bool Equals(object? obj)
    {
        if (obj is not LinqSelectFieldExpr other) return false;
        return Name == other.Name && Equals(Value, other.Value);
    }
    public override int GetHashCode() => (Name, Value).GetHashCode();
}